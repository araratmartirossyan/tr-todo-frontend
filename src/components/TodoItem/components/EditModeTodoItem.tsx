import { FC, FormEvent, useState, useRef, useEffect } from 'react'

// Components
import SaveIcon from '@/assets/icons/save.svg?component'
import { Button, StyledInput } from '@/components'

import { StyledTodoItem, ActionsBlock } from '../styles'

export interface EditModeProps {
  onSetEditMode: (editMode: boolean) => void
  onUpdate: (args: TODO.UpdateArguments) => void
  _id: string
  title: string
  done: boolean
}

export const EditModeTodoItem: FC<EditModeProps> = ({
  onUpdate,
  onSetEditMode,
  _id,
  title,
  done,
}) => {
  const [localTaskTitle, setLocalTaskTitle] = useState('')

  const inputRef =
    useRef<HTMLInputElement>() as React.MutableRefObject<HTMLInputElement>

  const onTaskInput = ({
    currentTarget: { value },
  }: FormEvent<HTMLInputElement>) => setLocalTaskTitle(value)

  useEffect(() => {
    setLocalTaskTitle(title)
    if (inputRef.current) {
      inputRef.current.focus()
    }
  }, [inputRef, title])

  const handleUpdate = () => {
    onSetEditMode(false)
    onUpdate({ title: localTaskTitle, _id, done })
  }

  return (
    <StyledTodoItem>
      <StyledInput
        className="todo-edit-input"
        name="title"
        data-test-id="todo-edit-input"
        onInput={onTaskInput}
        value={localTaskTitle}
        ref={inputRef}
      />

      <ActionsBlock>
        <Button
          tertiary
          className="todo-icon-button"
          data-test-id="todo-save-button"
          onClick={handleUpdate}
        >
          <SaveIcon />
        </Button>
      </ActionsBlock>
    </StyledTodoItem>
  )
}
