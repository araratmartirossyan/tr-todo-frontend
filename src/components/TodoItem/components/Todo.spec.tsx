import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { EditModeTodoItem, EditModeProps } from './EditModeTodoItem'

interface TestState extends EditModeProps {
  title: string
  done: boolean
  _id: string
  onSetEditMode: jest.Mock
  onUpdate: jest.Mock
}

describe('TodoItem test', () => {
  const defaultProps: TestState = {
    title: 'Test Task',
    done: false,
    _id: '11',
    onSetEditMode: jest.fn(),
    onUpdate: jest.fn(),
  }

  const createComponent = (testProps: TestState) =>
    shallow(<EditModeTodoItem {...testProps} />)

  it('should render component', () => {
    const component = createComponent(defaultProps)
    expect(toJson(component)).toMatchSnapshot()
  })

  it('on save button should fire updateEvent event', () => {
    const component = createComponent(defaultProps)
    const editBtn = component.find('[data-test-id="todo-save-button"]')
    expect(editBtn.exists()).toBe(true)
    editBtn.simulate('click')

    expect(defaultProps.onUpdate).toBeCalled()
  })

  it('on save button should fire updateEvent event', () => {
    const component = createComponent(defaultProps)
    const editBtn = component.find('[data-test-id="todo-save-button"]')
    expect(editBtn.exists()).toBe(true)
    editBtn.simulate('click')
    expect(defaultProps.onUpdate).toBeCalled()
    expect(defaultProps.onSetEditMode).toBeCalled()
  })
})
