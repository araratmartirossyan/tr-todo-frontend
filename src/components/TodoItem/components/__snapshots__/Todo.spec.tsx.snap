// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem test should render component 1`] = `
<styled.li>
  <styled.input
    className="todo-edit-input"
    data-test-id="todo-edit-input"
    name="title"
    onInput={[Function]}
    value=""
  />
  <styled.div>
    <Button
      className="todo-icon-button"
      data-test-id="todo-save-button"
      onClick={[Function]}
      tertiary={true}
    >
      <Component />
    </Button>
  </styled.div>
</styled.li>
`;
